drivers/counter/counter_nxp_pit.c

```diff
@@ -0,0 +1,407 @@
+{
+	const struct nxp_pit_config *config = dev->config;
+	struct nxp_pit_channel_data *data = PIT_CHANNEL_DATA(dev);
+	pit_chnl_t channel = PIT_CHANNEL_ID(dev);
+	uint32_t flags;
+
+	LOG_DBG("pit counter isr");
+
+	flags = PIT_GetStatusFlags(config->base, channel);
+	if (flags) {
+		PIT_ClearStatusFlags(config->base, channel, flags);
+		data->top_callback(dev, data->top_user_data);
+	}
+}
+#endif /* DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts) */
+
+static int nxp_pit_init(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	pit_config_t pit_config;
+	uint32_t clock_rate;
+
+	if (!device_is_ready(config->clock_dev)) {
+		LOG_ERR("Clock control device not ready");
+		return -ENODEV;
+	}
+
+	PIT_GetDefaultConfig(&pit_config);
+	pit_config.enableRunInDebug = config->enableRunInDebug;
+
+	PIT_Init(config->base, &pit_config);
+
+	clock_rate = nxp_pit_get_frequency(dev);
+
+#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
+	config->irq_config_func(dev);
+	for (int channel_index = 0;
+		channel_index < config->num_channels;
+		channel_index++) {
+		PIT_SetTimerPeriod(config->base, channel_index,
+			USEC_TO_COUNT(config->info.max_top_value, clock_rate));
+	}
+#else
+	for (int channel_index = 0;
+		channel_index < config->num_channels;
+		channel_index++) {
+		if (config->irq_config_func[channel_index]) {
+			config->irq_config_func[channel_index](dev);
+			PIT_SetTimerPeriod(config->base, channel_index,
+				USEC_TO_COUNT(config->info.max_top_value, clock_rate));
+		}
+	}
+#endif /* DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts) */
+	return 0;
+}
+
+static const struct counter_driver_api nxp_pit_driver_api = {
+	.start = nxp_pit_start,
+	.stop = nxp_pit_stop,
+	.get_value = nxp_pit_get_value,
+	.set_top_value = nxp_pit_set_top_value,
+	.get_pending_int = nxp_pit_get_pending_int,
+	.get_top_value = nxp_pit_get_top_value,
+	.get_freq = nxp_pit_get_frequency,
+};
+
+
+/* Creates a device for a channel (needed for counter API) */
+#define NXP_PIT_CHANNEL_DEV_INIT(node, pit_inst)						\
+	DEVICE_DT_DEFINE(node, NULL, NULL,							\
+		(void *)									\
+		&nxp_pit_##pit_inst##_channel_datas[DT_REG_ADDR(node)],				\
+		&nxp_pit_##pit_inst##_config,							\
+		POST_KERNEL, CONFIG_COUNTER_INIT_PRIORITY,					\
+		&nxp_pit_driver_api);
+
+/* Creates a decleration for each pit channel */
+#define NXP_PIT_CHANNEL_DECLARATIONS(node)  static struct nxp_pit_channel_data			\
+	nxp_pit_channel_data_##node;
+
+/* Initializes an element of the channel data pointer array */
+#define NXP_PIT_INSERT_CHANNEL_INTO_ARRAY(node)							\
+	[DT_REG_ADDR(node)] =									\
+		&nxp_pit_channel_data_##node,
+
+#define NXP_PIT_INSERT_CHANNEL_DEVICE_INTO_ARRAY(node)						\
+	[DT_REG_ADDR(node)] = DEVICE_DT_GET(node),
+
+#define NXP_PIT_INC_FOR_CHANNEL_COUNT(ignored) 1 +
+
+#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
```

Comment:
I would disagree on the macro becoming easier to read, here we have two init macros, one per compilation run and both kept in their own scope of the file, if I create only one init macro as you suggest, the two changes within the the init macro will be required to become submacros(config struct and the irq config definitions) as you say, I think this will look equally if not more complicated then the current design. The config struct as an example makes up a large portion of the Init macro, since one config contains an array of pointer where the other does not (.irq_config_func) this will have to become two sub macros one per compilation run. Secondly the irq macros that were specific to the second init macro now become specific to a unified init macros which makes less sense to someone new looking at the file.
If you still think the opposite I can make the change, just understand that the code will be equally difficult if not harder to read for newer readers of the file.

Reply:
I very much disagree but I'm not blocking over this since it might be personal preference

But one advantage to how I described is to have single point of control, since there is so much duplicated code between these two macros that is exactly the same, you wouldn't have to make changes in two places in the future

Reply:
resolved